selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+    link(inode_map[st[ST_INO]], target_obj)
+  else:
+    # Recreate all directories and symlinks; copy ownership and permissions.
+    if S_ISDIR(st[ST_MODE]):
+      os.mkdir(os.path.join(FINALIZED, source_obj), S_IMODE(st[ST_MODE]))
+    elif S_ISLNK(st[ST_MODE]):
+      os.symlink(os.readlink(source_obj), target_obj)
+      os.chmod(target_obj, S_IMODE(st[ST_MODE]), follow_symlinks=False)
+    else:
+      # Ran into a FIFO, socket, etc. Should not happen in OP install dir.
+      # Ignore without copying for the time being, revisit if necessary.
+      pass
+    os.chown(target_obj, st[ST_UID], st[ST_GID], follow_symlinks=False)
+    os.utime(target_obj, (st[ST_ATIME], st[ST_MTIME]), follow_symlinks=False)
+
+def finalize_from_ovfs():
+  # Take the current OverlayFS merged view and finalize a copy outside of
+  # OverlayFS, ready to be swapped-in at BASEDIR.
+  #
+  # The "copy" is done with hardlinks, but since the OverlayFS merge looks
+  # like a different filesystem, and hardlinks can't cross filesystems, we
+  # have to link from a pathname in the upper or lower layers.
 
-    # Write latest release notes to param
+  cloudlog.info("creating overlay inode index")
+  inode_map = inodes_in_tree(BASEDIR)
+  inode_map.update(inodes_in_tree(OVERLAY_UPPER))
+
+  cloudlog.info("creating finalized version of the overlay")
+  shutil.rmtree(FINALIZED)
+  os.umask(0o077)
+  os.mkdir(FINALIZED)
+  os.chdir(OVERLAY_MERGED)
+  for root, dirs, files in os.walk('.', topdown=True):
+    for obj_name in dirs:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+    for obj_name in files:
+      dup_ovfs_object(inode_map, os.path.join(root, obj_name), FINALIZED)
+  os.chdir(BASEDIR)  # otherwise umount will fail later with "target is busy"
+
+def attempt_update():
+  upd_lock_fd = open(UPDATE_LOCK, 'r')
+  try:
+    fcntl.flock(upd_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+  except OSError:
+    print("couldn't get update lock; is another updated running? going back to sleep...")
+    upd_lock_fd.close()
+    return
+
+  cloudlog.info("attempting git update inside staging overlay")
+  os.chdir(OVERLAY_MERGED)
+
+  # Un-set the successful flag to prevent the finalized tree from being
+  # activated later if the update fails in some way.
+  if os.path.isfile(os.path.join(FINALIZED, ".update_succeeded")):
+    os.remove(os.path.join(FINALIZED, ".update_succeeded"))
+  os.system("sync")
+
+  r = run(NICE_LOW_PRIORITY + ["git", "fetch"], OVERLAY_MERGED)
+  cloudlog.info("git fetch success: %s", r)
+
+  cur_hash = run(["git", "rev-parse", "HEAD"], OVERLAY_MERGED).rstrip()
+  upstream_hash = run(["git", "rev-parse", "@{u}"], OVERLAY_MERGED).rstrip()
+
+  cloudlog.info("comparing %s to %s" % (cur_hash, upstream_hash))
+  if cur_hash != upstream_hash:
+    cloudlog.info("git reset in progress")
+    r = [run(NICE_LOW_PRIORITY + ["git", "reset", "--hard", "@{u}"], OVERLAY_MERGED),
```

Comment:
should we check the exit codes of these calls?

Reply:
Generally yes; I intend to do so. I have some experimentation to do, deliberately corrupting the Git repo by various means, to see how I can best detect that result from Git during updated runtime. I may or may not be able to get that from numeric exit code, I may need to parse its output. For example, I may not be able to distinguish git fetch failure due to poor Internet connectivity (recoverable) from failure due to Git repo corruption (requires active measures to reinit Git).

Reply:
Well, we never have to nuke the git repo, right? Can't we always assume the git repo in the BASEDIR is correct? We just have to start with a new overlay if any of these commands fail.

Reply:
> Can't we always assume the git repo in the BASEDIR is correct?

It _should_ be, but geohot had a strong opinion about keeping the ability to develop directly on-EON, and if that's happening:
- BASEDIR .git is now re-exposed to the corruption scenarios updated is designed to protect it from
- I have to watch and plan for BASEDIR .git to be messed with while updated is running (have not added the check for this in my code yet)